vegeta_wrapper/trigger_vegeta.py

```diff
@@ -0,0 +1,106 @@
+
+    def _json_payload(self, data, sample):
+        payload = {
+            "workload": "vegeta",
+            "uuid": self.uuid,
+            "user": self.user,
+            "cluster_name": self.cluster_name,
+            "iteration": sample,
+            "duration": self.duration,
+            "workers": self.workers,
+            "keepalive": self.keepalive,
+        }
+        payload.update(data)
+        return payload
+
+    def _run_vegeta(self):
+        cmd = (
+            "vegeta attack -keepalive={0} -insecure -workers={1} -duration={2}s -targets={3} -rate=0"
+            " -max-workers={1} | vegeta report --every=1s --type=json --output=vegeta.log").format(
+            self.keepalive, self.workers, self.duration, self.targets)
+        logger.info(cmd)
+        p = subprocess.run(cmd, shell=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        return p.stdout.strip().decode("utf-8"), p.stderr.strip().decode("utf-8"), p.returncode
+
+    def _parse_stdout(self):
+        status_codes = {}
+        status_codes_bck = {}
+        bytes_in_bck = 0
+        bytes_out_bck = 0
+        for l in open("vegeta.log").readlines():
+            data = json.loads(l)
+            rps = int(data["rate"])
+            throughput = int(data["throughput"])
+            for s, n in data["status_codes"].items():
+                if s in status_codes_bck:
+                    status_codes[s] = n - status_codes_bck[s]
+                else:
+                    status_codes[s] = n
+                status_codes_bck[s] = n
+            bytes_in = data["bytes_in"]["total"] - bytes_in_bck
+            bytes_in_bck = data["bytes_in"]["total"]
+            bytes_out = data["bytes_out"]["total"] - bytes_out_bck
+            bytes_out_bck = data["bytes_out"]["total"]
+            # Latency units are reported in nanoseconds. We convert them here to usecs
+            p99 = int(data["latencies"]["99th"] / 1000)
+            ltcy = int(data["latencies"]["mean"] / 1000)
+            ts = dateutil.parser.parse(data["end"])
+            yield {
+                "rps": rps,
+                "throughput": throughput,
+                "status_codes": status_codes,
+                "requests": data["requests"],
+                "p99_latency": p99,
+                "req_latency": ltcy,
+                "timestamp": ts,
+                "bytes_in": bytes_in,
+                "bytes_out": bytes_out
+            }
+
+    def emit_actions(self):
+        if not os.path.exists(self.targets):
+            logger.critical("Targets file %s not found" % self.targets)
+            exit(1)
+        for s in range(1, self.sample + 1):
+            logger.info("Starting vegeta sample %d out of %d with uuid %s" % (s, self.sample, self.uuid))
+            stdout, stderr, rc = self._run_vegeta()
+            if rc:
+                logger.critical("Vegeta failed to execute, stopping")
+                logger.error("stderr: %s" % stderr)
```

Comment:
want to include `stdout: %s` too... Just in case vegeta provides useful information in stdout as well?

Reply:
fixed. Nice catch